### PR TITLE
Add support for manual entry

### LINKED ISSUE #37: Add support to enter word and count manually
* Users should be able to create a 'Direct Entry' source.

* When users select 'Direct Entry' as their source type, a table should appear with three columns (word, word count, actions) and one 'new row'.

* Each row will have one input field for the word, one optional input field for the word count, and one 'delete' icon.

  * Users can edit values for existing rows by selecting the input field.

  * When users select a row, the styling of the row should change to indicate it has been selected.

  * Users can type in a word into a new row and when they do another new row should appear below automatically.

  * Words have to be strings and cannot be empty.

  * Word counts have to be > 0 or empty.

  * Clicking the delete icon will remove the appropriate row from the table.

 * Users can enter a name for the Direct Entry Table via a text input field.

* The table will automatically save periodically and will have a manual save button.

* The table has to have at least one valid row to be saved or added.

### PR DESCRIPTION
**Pull request by xqiu1**
Fixes #37

### CODE REVIEW

File: src/app/components/ManualEntry.svelte [hunk 1]
```diff
@@ -1,9 +1,57 @@
 <script lang="ts">
-  /**
-   * Handles the click when the delete button is pressed
-   * TODO
-   */
-  const deleteRow = (): void => {};
+  import Button from "./Button.svelte";
+  import worker from "../spawn-worker";
+
+  interface rowDataObj {
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte]: Interface names should be in PascalCase and should avoid the words `obj` and `interface` in the name. How about `DictionaryEntry`?

```suggestion
  interface DictionaryEntry {
```

(See: https://github.com/airbnb/javascript#naming--PascalCase)

File: src/app/components/ManualEntry.svelte [hunk 2]
```diff
@@ -1,9 +1,57 @@
...
+    word: string;
+    count: number | undefined;
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte]:
```suggestion
    count?: number;
```

File: src/app/components/ManualEntry.svelte [hunk 3]
```diff
@@ -1,9 +1,57 @@
...
+    },
+  ];
+  $: validInput = rowDataFromManualEntry.every((element) =>
+    validateRowData(element)
+  );
+  $: btnColor = validInput ? "blue" : "grey";
+
+  const validateRowData = (rowData: rowDataObj): boolean => {
+    const word = rowData.word;
+    return word !== "" && word !== undefined && word !== null;
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte]: Please add a check that word is not just whitespace! e.g., `"  \t \n"` is not a word in any language!

File: src/app/components/ManualEntry.svelte [hunk 4]
```diff
@@ -42,39 +83,46 @@
...
+  table tr:last-of-type td:first-of-type {
     border-bottom-left-radius: 10px;
   }
-  tr:last-of-type td:last-of-type {
+  table tr:last-of-type td:last-of-type {
     border-bottom-right-radius: 10px;
   }
 
-  .btn--inline {
+  table .btn--inline {
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte]: Svelte has "scoped styles" so these styles only apply to elements in _this_ `.svelte` component. Why did you add the extra `table ` to all these selectors? Especially since `td`, `th` and `tr` _must_ be children of a `<table>` element.

File: src/app/components/ManualEntry.svelte [hunk 5]
```diff
@@ -42,39 +83,46 @@
...
+  table tr {
     border-bottom: 1pt solid var(--lite-white);
   }
 
-  td {
+  table td {
     vertical-align: middle;
     padding: 0.625rem 0 0.625rem 1.25rem;
     color: var(--gray-dark);
     text-align: center;
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte]: I think center alignment should be avoided. [See this answer on UX stack exchange](https://ux.stackexchange.com/a/113594). The correct alignment depends on the nature of the content, but I think for words it should be left aligned for LTR langauges (e.g., English, Chinese) and right-aligned for RTL languages (e.g., Arabic Hebrew). You can use the `text-align: start` for this:

```suggestion
    text-align: start;
```

Arabic numbers (i.e., 1234567890) should _always_ be right-aligned.

File: src/app/components/ManualEntry.svelte [hunk 6]
```diff
@@ -42,39 +83,46 @@
...
     display: inline;
     padding: 0;
     border: 0;
     background-color: transparent;
     font-size: 1em;
   }
 
-  td > input {
+  table td > input {
     height: 20px;
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte]: Use `rem` units for this measure, so that it scales with the containing font-size

File: src/app/components/ManualEntry.svelte [hunk 7]
```diff
@@ -84,50 +132,87 @@
     font: 1em var(--main-font);
   }
 
-  input:focus {
+  table input:focus {
     background-color: var(--lite-white);
   }
+
+  table p {
+    display: inline;
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte]: I strongly advice against setting `p` to have `display: inline`. `<p>` is semantically a block-level element, so the HTML spec actually treats it as a "block" even if it appears inline.

File: src/app/components/ManualEntry.svelte [hunk 8]
```diff
@@ -84,50 +132,87 @@
...
+    {#each rowDataFromManualEntry as row, i (i)}
+      <tr>
+        <td>
+          <input
+            type="text"
+            bind:value={row.word}
+            data-cy="manual-entry-input-word" />
+          {#if !validateRowData(row)}
+            <p>*</p>
+          {/if}
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte]: Some of the validation can be done automatically by the browser. If you want to make sure the field is filled, use [the `required` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/required). No JavaScript needed!

```suggestion
          <input
            type="text"
            bind:value={row.word}
            required
            data-cy="manual-entry-input-word" />
```

I would not keep the \*, but if you really want it, wrap it in a `<span>`; not a `<p>`

File: src/app/components/ManualEntry.svelte [hunk 9]
```diff
@@ -84,50 +132,87 @@
...
-          data-cy="manual-entry-input-count" />
-      </td>
-      <td>
+      <td colspan="3">
         <button
           class="btn--inline"
-          on:click={deleteRow}
-          data-cy="manual-entry-delete">Delete</button>
+          on:click={addNewRow}
+          data-cy="munual-entry-add-row">Add Row</button>
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte]: Typo:

```suggestion
          data-cy="manual-entry-add-row">Add Row</button>
```

File: src/app/components/ManualEntry.svelte [hunk 10]
```diff
@@ -1,9 +1,57 @@
...
+    const word = rowData.word;
+    return word !== "" && word !== undefined && word !== null;
+  };
+
+  const deleteRow = (i: number): void => {
+    const part = rowDataFromManualEntry.splice(i, 1);
+    rowDataFromManualEntry = rowDataFromManualEntry;
+  };
+
+  const addNewRow = (): void => {
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte:49]: I really like the fact that this is called `addNewRow` and not like... `handleClick`. It makes sense EXACTLY what this will do when clicked!

File: src/app/components/ManualEntry.svelte [hunk 11]
```diff
@@ -84,50 +132,87 @@
...
-                <button class="btn--inline" on:click={deleteRow}>Delete</button>
-            </td>
-        </tr>
   </table>
+
+  <div class="save-zone">
+    <Button
+      color={btnColor}
+      isOutlined
+      disabled={!validInput}
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte]: I do not like this solution (having a manual save button), as I think an auto-save option is better. However, I think autosave is enough work that it's a new PR 😅 

The problem with a button that gets disabled on invalid input is that there's no written feedback as to what exactly is wrong with the form. "Why can't I save"? The `required` attribute of `<input>` would help with this. Note: you may want to wrap the entire table in `<form>` element, and then make the underlying button here be a `<button type="submit">`. This will enable the browser to automatically validate the `required` attribute AND tell you which words are missing!

File: src/common/predictive-text-studio-worker.ts [hunk 12]
```diff
@@ -12,13 +12,21 @@ export interface PredictiveTextStudioWorker {
...
   addDictionarySourceToProject(name: string, contents: File): Promise<number>;
 
+  /**
+   * Store the manual entry data into the database
+   * @param tableData Manual entry data
+   */
+  addManualEntryDictionaryToProject(tableData: {
+    name: string;
+    data: { word: string; count: number | undefined }[];
+  }): Promise<number>;
```
Comment (eddieantonio) [inline src/common/predictive-text-studio-worker.ts]: Could you please extract an interface for this type?

File: src/worker/read-wordlist.ts [hunk 13]
```diff
@@ -45,3 +45,15 @@ function asNonNegativeInteger(x: unknown): number {
 
   return n;
 }
+
+export async function readManualEntryData(
+  contents: { word: string; count: number | undefined }[]
+): Promise<WordList> {
```
Comment (eddieantonio) [inline src/worker/read-wordlist.ts]: Only functions that actually use promises or other the `await` syntax should be marked as `async`:

```suggestion
export function readManualEntryData(
  contents: { word: string; count: number | undefined }[]
): WordList {
```

File: src/worker/read-wordlist.ts [hunk 14]
```diff
@@ -45,3 +45,15 @@ function asNonNegativeInteger(x: unknown): number {
...
+  const wordlist: WordList = [];
+  contents.forEach((row) => {
+    const word = row.word;
+    const count = asNonNegativeInteger(row.count || 0);
+    wordlist.push([word, count]);
+  });
+  return wordlist;
```
Comment (eddieantonio) [inline src/worker/read-wordlist.ts]: This is a good candidate for `Array.prototype.map()`:

```suggestion
  const wordlist: WordList = contents.map((row) => {
    const word = row.word;
    const count = asNonNegativeInteger(row.count || 0);
    return [word, count];
  });
  return wordlist;
```

**Review by eddieantonio — CHANGES_REQUESTED** [review]
I have a LOT of nit picky feedback. Please feel free to DM me if you have any questions!

File: src/app/components/ManualEntry.svelte (hunk 1 above)
Comment (xqiu1) [reply src/app/components/ManualEntry.svelte]: Sure, it makes more sense

File: src/app/components/ManualEntry.svelte [hunk 15]
```diff
@@ -1,9 +1,80 @@
...
+      name: "action",
+    },
+  ];
+  $: rowDataFromManualEntry = tableData.data;
+  $: validDictionary =
+    validateRowsData(rowDataFromManualEntry) &&
+    validateTableName(tableData.name);
+  $: if (validDictionary) {
+    worker.addManualEntryDictionaryToProject(tableData);
+  }
```
Comment (xqiu1) [inline src/app/components/ManualEntry.svelte]: I add the basic support for autosaving dictionary.

File: src/app/components/Button.svelte [hunk 16]
```diff
@@ -28,6 +29,7 @@
     class="button button--{color} button--{size}"
     class:button--shadow={hasDropShadow}
     class:button--outline={isOutlined}
+    {type}
```
Comment (eddieantonio) [inline src/app/components/Button.svelte:32]: This is a great change! I found out recently that [the default type of a button is `type="submit"`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#attr-type) which... is not always helpful 😅

File: src/app/components/ManualEntry.svelte [hunk 17]
```diff
@@ -1,9 +1,80 @@
...
+    },
+    {
+      label: "Count",
+      name: "count",
+    },
+    {
+      label: "Action",
+      name: "action",
+    },
+  ];
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte]: Honestly, I'd rather these columns be hardcoded in the HTML below for the time being rather than using this array. I'm not sure what benefit this array brings.

File: src/app/components/ManualEntry.svelte [hunk 18]
```diff
@@ -1,9 +1,80 @@
...
+
+  const validateRowsData = (rowsData: DictionaryEntry[]): boolean => {
+    return rowsData.every((rowData) => {
+      const word = rowData.word;
+      return (
+        word !== "" &&
+        word !== undefined &&
+        word !== null &&
+        !word.match(/^([\s\t\r\n]*)$/)
+      );
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte]: I feel like there's some code duplication that could be factored out here! In another PR though!

File: src/app/components/ManualEntry.svelte [hunk 19]
```diff
@@ -1,9 +1,80 @@
...
+
+  const saveTableData = async () => {
+    const validDictionary: boolean =
+      validateRowsData(rowDataFromManualEntry) &&
+      validateTableName(tableData.name);
+    if (validDictionary) {
+      const numOfWordStored = await worker.addManualEntryDictionaryToProject(
+        tableData
+      );
+    }
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte]: There's some duplicated code here with the reactive statementsthat handle autosave (`$: if (validDictionary) ...`). Can you reduce the duplication here?

File: src/app/components/ManualEntry.svelte [hunk 20]
```diff
@@ -87,47 +159,78 @@
...
+    display: flex;
+    flex-direction: row;
+    align-items: center;
+    justify-content: center;
+    padding: 1%;
+  }
 </style>
 
-<div class="language__sources-manual-entry">
+<form class="language__sources-manual-entry" on:submit|preventDefault>
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte:155]: 👍🏼

File: src/app/components/ManualEntry.svelte [hunk 21]
```diff
@@ -87,47 +159,78 @@
...
-  <table>
+  <table id="manual-entry-table">
     <thead>
-      <th>Word</th>
-      <th>Count</th>
-      <th>Action</th>
+      {#each columns as column (column.name)}
+        <th>{column.label}</th>
+      {/each}
+      <th />
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte]: Was this left here by mistake?

```suggestion
```

File: src/app/components/ManualEntry.svelte [hunk 22]
```diff
@@ -87,47 +159,78 @@
...
 
-  <table>
+  <table id="manual-entry-table">
     <thead>
-      <th>Word</th>
-      <th>Count</th>
-      <th>Action</th>
+      {#each columns as column (column.name)}
+        <th>{column.label}</th>
+      {/each}
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte]: This is what I meant. I think it's best to just hard code the headers here.
```suggestion
      <th> Word </th>
      <th> Count </th>
      <th> Action </th>
```

File: src/app/components/ManualEntry.svelte [hunk 23]
```diff
@@ -87,47 +159,78 @@
...
+            type="number"
+            min="0"
+            placeholder="Optional"
+            bind:value={row.count}
+            data-cy="manual-entry-input-count" />
+        </td>
+        <td>
+          <button
+            class="btn--inline"
+            on:click={() => deleteRow(i)}
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte:192]: very nice!

File: src/app/components/ManualEntry.svelte [hunk 24]
```diff
@@ -87,47 +159,78 @@
...
-                <button class="btn--inline" on:click={deleteRow}>Delete</button>
-            </td>
-        </tr>
   </table>
-</div>
+
+  <div class="save-zone">
+    <Button
+      type={'submit'}
+      color={'blue'}
```
Comment (eddieantonio) [inline src/app/components/ManualEntry.svelte]:
```suggestion
      type="submit"
      color="blue"
```

File: src/common/manual-entry-file.ts [hunk 25]
```diff
@@ -0,0 +1,4 @@
+export interface DictionaryEntry {
```
Comment (eddieantonio) [inline src/common/manual-entry-file.ts]: I like the new interface, but could you please put it in in `src/common/types.ts`? I don't want there to be an explosion of files. Not every interface needs its own source file!

Also, please add JSDoc comments!

File: src/worker/predictive-text-studio-worker-impl.ts [hunk 26]
```diff
@@ -87,6 +88,16 @@ export class PredictiveTextStudioWorkerImpl
     return wordlist.length;
   }
 
+  async addManualEntryDictionaryToProject(tableData: {
+    name: string;
+    data: DictionaryEntry[];
+  }): Promise<number> {
+    const dictionaryName = tableData.name;
+    const wordlist = await readManualEntryData(tableData.data);
```
Comment (eddieantonio) [inline src/worker/predictive-text-studio-worker-impl.ts]:
```suggestion
    const wordlist = readManualEntryData(tableData.data);
```

File: src/worker/read-wordlist.ts [hunk 27]
```diff
@@ -45,3 +45,12 @@ function asNonNegativeInteger(x: unknown): number {
 
   return n;
 }
+
+export function readManualEntryData(contents: DictionaryEntry[]): WordList {
```
Comment (eddieantonio) [inline src/worker/read-wordlist.ts:49]: 👍🏼

**Review by eddieantonio — CHANGES_REQUESTED** [review]
Looking better! There are still a few things to polish up!

File: src/app/components/ManualEntry.svelte (hunk 19 above)
Comment (xqiu1) [reply src/app/components/ManualEntry.svelte]: Sure, I will try to factor out the duplication 😅

**Review by eddieantonio — APPROVED** [review]
LGTM!
